presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -1220,15 +1222,28 @@ private void commitShared()
         private final List<AlterPartitionOperation> alterPartitionOperations = new ArrayList<>();
         private final List<UpdateStatisticsOperation> updateStatisticsOperations = new ArrayList<>();
         private final List<IrreversibleMetastoreOperation> metastoreDeleteOperations = new ArrayList<>();
+        private final List<TransactionalMetastoreOperation> metastoreTransactionalOperations = new ArrayList<>();
 
         // Flag for better error message
         private boolean deleteOnly = true;
 
-        private void prepareDropTable(HiveIdentity identity, SchemaTableName schemaTableName)
+        private void prepareDropTable(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName)
         {
-            metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
-                    format("drop table %s", schemaTableName),
-                    () -> delegate.dropTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName(), true)));
+            Table table = delegate.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName())
+                    .orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            if (AcidUtils.isTransactionalTable(table.getParameters())) {
```

Comment:
It feel that you can just wrap the operation into locking logic and use `metastoreDeleteOperations` list. Here.

Sth like this:
```java
    public Runnable runWithLock(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName, Runnable action)
    {
        return () -> {
            long heartbeatInterval = configuredTransactionHeartbeatInterval
                    .map(Duration::toMillis)
                    .orElseGet(SemiTransactionalHiveMetastore.this::getServerExpectedHeartbeatIntervalMillis);
            long transactionId = delegate.openTransaction(identity);

            String queryId = context.getQueryId().orElseThrow(() -> new IllegalArgumentException("query ID not present"));
            log.debug(format("Opened transaction: %s for queryID: %s for table: %s", transactionId, queryId, schemaTableName.getTableName()));

            ScheduledFuture<?> heartbeatTask = heartbeatExecutor.scheduleAtFixedRate(
                    () -> delegate.sendTransactionHeartbeat(identity, transactionId),
                    0,
                    heartbeatInterval,
                    MILLISECONDS);

            delegate.acquireLock(
                    identity,
                    queryId,
                    transactionId,
                    ImmutableList.of(schemaTableName),
                    ImmutableList.of(),
                    DataOperationType.NO_TXN);

            try {
                action.run();
            }
            finally {
                heartbeatTask.cancel(true);
                delegate.commitTransaction(identity, transactionId);
            }
        };
    }
```

and then
```java
        private void prepareDropTable(HdfsContext context, HiveIdentity identity, SchemaTableName schemaTableName)
        {
            Table table = delegate.getTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName())
                    .orElseThrow(() -> new TableNotFoundException(schemaTableName));

            Runnable operation = () -> delegate.dropTable(identity, schemaTableName.getSchemaName(), schemaTableName.getTableName(), true);
            if (AcidUtils.isTransactionalTable(table.getParameters())) {
                operation = runWithLock(context, identity, schemaTableName, operation);
            }

            metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
                    format("drop table %s", schemaTableName),
                    operation));
        }
```

Reply:
The notion of creating `TransactionalMetastoreOperation` is to register actions which we want to run on a transactional table by acquiring a lock and then run the action. Right now, change is only for `drop table`, but we may be needing this for other DDL commands like `alter table` which need an exclusive lock.

The way `run` method is structured, keeping this into an account. We can register any transactional table action with it and `run` will take care of opening a transaction, creating heartbeat, acquiring lock, commit/rollingback transaction.